File: 2023/07/ts/main.ts
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

export const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

const ranking: Record<string, number> = {
   A: 13,
   K: 12,
   Q: 11,
   J: 10,
   T: 9,
   '9': 8,
   '8': 7,
   '7': 6,
   '6': 5,
   '5': 4,
   '4': 3,
   '3': 2,
   '2': 1,
   '1': 0,
};

const enum Type {
   HIGH_CARD,
   ONE_PAIR,
   TWO_PAIR,
   THREE_OF_A_KIND,
   FULL_HOUSE,
   FOUR_OF_A_KIND,
   FIVE_OF_A_KIND,
}

function sortCard(a: [string, number], b: [string, number]): number {
   for (let i = 0; i < 5; i++) {
      if (a[0][i] !== b[0][i]) return ranking[a[0][i]] - ranking[b[0][i]];
   }
   return 0;
}

function getType(str: string): number {
   let values = new Array(14).fill(0);
   for (const k of str) values[ranking[k]]++;
   if (values[0]) {
      const temp = values[0];
      values[0] = 0;
      const idx = values.findIndex((e) => e === Math.max(...values));
      values[idx] += temp;
   }

   values = values.filter((e) => e);
   if (values.length === 1) return Type.FIVE_OF_A_KIND;
   if (values.length === 4) return Type.ONE_PAIR;
   if (values.length === 5) return Type.HIGH_CARD;

   let min = 5;
   let max = 0;
   for (const v of values) {
      if (min > v) min = v;
      if (max < v) max = v;
   }
   if (min === 1) {
      if (max === 2) return Type.TWO_PAIR;
      if (max === 3) return Type.THREE_OF_A_KIND;
      if (max === 4) return Type.FOUR_OF_A_KIND;
   }
   return Type.FULL_HOUSE;
}

function parseInput(input: string, joker: boolean): [string, number][][] {
   return input
      .split('\n')
      .map((str) => {
         const temp: unknown[] = str.split(' ');
         temp[1] = Number(temp[1]);
         if (joker) temp[0] = (temp[0] as string).replaceAll('J', '1');
         return temp as [string, number];
      })
      .reduce((p, set) => {
         const t = getType(set[0]);
         p[t] ||= [];
         p[getType(set[0])].push(set);
         return p;
      }, [] as [string, number][][]);
}

function solve(input: string, joker: boolean): string {
   const groups: [string, number][][] = parseInput(input, joker);

   let res = 0;
   let i = 1;
   for (let group of groups) {
      group = (group || []).sort(sortCard);
      for (const set of group) res += set[1] * i++;
   }
   return res.toString();
}

export function part1(input: string, _isTest: boolean): string {
   return solve(input, false);
}

export function part2(input: string, _isTest: boolean): string {
   return solve(input, true);
}

if (import.meta.main) {
   run(Deno.args, part1, part2, options);
}

File: 2023/07/lua/main.lua
local run = require('src.lua.run')
local str = require('src.lua.utils.str')
local array = require('src.lua.utils.array')

local options = {
   has_alternate = false,
   has_io = false,
}

local ranking = {
   ['A'] = 14,
   ['K'] = 13,
   ['Q'] = 12,
   ['J'] = 11,
   ['T'] = 10,
   ['9'] = 9,
   ['8'] = 8,
   ['7'] = 7,
   ['6'] = 6,
   ['5'] = 5,
   ['4'] = 4,
   ['3'] = 3,
   ['2'] = 2,
   ['1'] = 1,
}

local card_type = {
   ['HIGH_CARD'] = 1,
   ['ONE_PAIR'] = 2,
   ['TWO_PAIR'] = 3,
   ['THREE_OF_A_KIND'] = 4,
   ['FULL_HOUSE'] = 5,
   ['FOUR_OF_A_KIND'] = 6,
   ['FIVE_OF_A_KIND'] = 7,
}

local function sort_card(a, b)
   for i = 1, 5 do
      local x = a.cards:sub(i, i)
      local y = b.cards:sub(i, i)
      if x ~= y then
         return ranking[x] < ranking[y]
      end
   end
end

local function get_type(cards)
   local values = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
   for i = 1, 5 do
      local c = cards:sub(i, i)
      values[ranking[c]] = values[ranking[c]] + 1
   end
   if values[1] > 0 then
      local temp = values[1]
      values[1] = 0
      local idx = array.index_of(values, array.max(values))
      values[idx] = values[idx] + temp
   end

   values = array.filter(values, function(v) return v > 0 end)
   if #values == 1 then return card_type['FIVE_OF_A_KIND'] end
   if #values == 4 then return card_type['ONE_PAIR'] end
   if #values == 5 then return card_type['HIGH_CARD'] end

   local mn = 5
   local mx = 0
   for _, v in ipairs(values) do
      if mn > v then
         mn = v
      end
      if mx < v then
         mx = v
      end
   end

   if mn == 1 then
      if mx == 4 then
         return card_type['FOUR_OF_A_KIND']
      end
      if mx == 3 then
         return card_type['THREE_OF_A_KIND']
      end
      if mx == 2 then
         return card_type['TWO_PAIR']
      end
   end
   return card_type['FULL_HOUSE']
end

local function parse_input(input, joker)
   local groups = {
      {},
      {},
      {},
      {},
      {},
      {},
      {},
   }

   for line in input:gmatch('[^\n]+') do
      local temp = str.split(line, ' ')
      if joker then
         temp[1] = temp[1]:gsub('J', '1')
      end
      local p = {
         cards = temp[1],
         value = tonumber(temp[2]),
      }
      local t = get_type(temp[1])
      groups[t][#groups[t] + 1] = p
   end

   return groups
end

local function solve(input, joker)
   local groups = parse_input(input, joker)
   local res = 0
   local i = 1

   for _, group in ipairs(groups) do
      table.sort(group, sort_card)
      for _, pair in ipairs(group) do
         res = res + pair.value * i
         i = i + 1
      end
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   return solve(input, false)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   return solve(input, true)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()
